### ISSUE TITLE
 LiteX-boards support

### ISSUE THREAD

**Issue by tcal-x** [tinyvision-ai-inc/UPduino-v3.0#16]
It would be great to have support (both a platform file and a target file) in [litex-boards](https://github.com/litex-hub/litex-boards/tree/master/litex_boards/targets) for v3.1.   

I did find these for v1: https://github.com/timvideos/litex-buildenv/blob/master/targets/upduino_v1/base.py  (target file), https://github.com/timvideos/litex-buildenv/blob/master/platforms/upduino_v1.py (platform file).

Probably some of the pinout in the platform file would need updating, and it might need some massaging to move it to `litex-boards`.

**Comment by vr2045** [comment]
Good point. Will add to the list of todo's :)

**Comment by Xenador77** [comment]
Done with the GPIO pins (the easy part) just figuring out how to do the ftdi serial/spi-flash pins and the 12MHz clock.

**Comment by Xenador77** [comment]
https://github.com/tinyvision-ai-inc/litex-boards
Any/all feedback is appreciated, but I have (finally) managed to get the upduino_v3 platform/target files to work with the actual upduino_v3 hardware

**Comment by tcal-x** [comment]
Hi @Xenador77 , I have a v3.1 board, with pmod connector, with no modifications.    Should I expect these board files to work for me?   I tried adding LedChaser (yes even though there's just the single RGB LED), since that is usually the first sign of life of LiteX running on the board.   But I don't see any activity.   Is there a reset signal I need to toggle or something like that?

I know the build is wiring to the LED correctly, since my first attempt I used polarity=0 in the LED chaser, and the LED was just bright white.   Then I used polarity=1, and the LED is off.

**Comment by Xenador77** [comment]
The files **should** work for both 3.0 and 3.1 (I've only tested with the 3.1). However, there is a small modification needed (oops):
The jumper labeled "OSC" needs to be solder bridged or gpio_20 (pin 44) needs to be connected to the 12MHz pin (pin 41)  
@XarkLabs made a useful ASCII pinout diagram that better illustrates the location of the pins/jumper
![image](https://user-images.githubusercontent.com/69032898/181882465-0199f795-3963-48bb-8d15-0a0d08680243.png)

I initially included the LedChaser in the target file, but removed it due to the brightness (although my RGB LED was pinkish-red)

**Comment by tcal-x** [comment]
@Xenador77 , yeah, I agree the LED is so bright.  And you're right it's not pure white.   Quickfeather has the same blinding brightness issue.  

The jumper should be easy, since I'm already using the pin header on that side (for the uart).   I just need to solder those pins.   I'll let you know how it goes.

**Comment by tcal-x** [comment]
@Xenador77 , no success yet at my end.   What was the exact LiteX recipe you used?   I had tried:
```
./upduino_v3.py --build --cpu-type=vexriscv --cpu-variant=minimal --flash
```

**Comment by tcal-x** [comment]
@Xenador77 , I needed to make this fix; now I see (the RGB version of) LedChaser working :)

```diff
 _io = [
     # CLK / RST
-    ("clk12", 0, Pins("35"), IOStandard("LVCMOS33")),
+    ("clk12", 0, Pins("20"), IOStandard("LVCMOS33")),

```

**Comment by tcal-x** [comment]
@Xenador77 I'm not having any luck with the UART/TTY though; did it work for you?

**Comment by Xenador77** [comment]
I'm working on figuring out what is going on with that right now
Just to confirm, you are using an external UART board on pins 2 and 3? 
(ie FT232 or FT2232 etc. anything that does UART at 3.3v)
It may have something to do with  https://github.com/tinyvision-ai-inc/UPduino-v3.0/issues/13#issue-873805660

**Comment by tcal-x** [comment]
Correct, I'm using the Digilent Pmod USBUART, but not in the pmod header, instead directly connected using jumper wires to pins 2, 3, and GND of the Upduino board.   I had tried a different usb/uart board beforehand also with no luck, but I switched to the Digilent board because it has tx/rx LEDs.   I'm not getting any data coming from the Upduino, but that could be due to anything preventing the CPU from working correctly (SPIflash etc).

I guess we can look at the schematics for Fomu and/or Icebreaker, which can read from spiflash.

**Comment by Xenador77** [comment]
No luck yet on my end, tried adding
 `Misc("PULLUP")`
to the tx pin, nothing seemed to change.
I'm going to try dumping the flash to see if everything is actually writing correctly to it

**Comment by tcal-x** [comment]
The Upduino schematic looks pretty much the same as Fomu's w.r.t. to the spiflash connection.   

https://github.com/im-tomu/fomu-hardware/blob/master/archive/pvt/releases/pvt1/tomu-fpga-pvt1.pdf

**Comment by jonnor** [comment]
Has progress been made on the UART not working on UPDuino with LiteX? And what is the status of your fork in general @Xenador77  - looking at the code it looks reasonably complete? Perhaps it can be upstreamed?

And I assume that the OSC pin changes mentioned above is still needed?

EDIT: for reference, the latest fork for UPduino v3, seems to be at https://github.com/Xenador77/litex-boards/blob/master/litex_boards/targets/upduino_v3.py (not the tinyvision-ai repo linked above)

**Comment by jonnor** [comment]
The xenador77 fork seems to be for litex 2025.04. However when I built with that version, I got several failures.
I managed to fix these in https://github.com/jonnor/litex-boards/tree/upduino_v3

However, I do not see any LiteX BIOS message on the UART. Nor does there seem to be any "blink" etc for this definition, so I do not really know if it is working or not.
